rflx/parser/grammar.py

```diff
@@ -429,93 +713,151 @@ def parse_array_aggregate(string: str, location: int, tokens: ParseResults) -> E
+
+    if tokens[0] == "all":
+        return ForAllIn(tokens[1], tokens[2], tokens[3], location=locn)
+    if tokens[0] == "some":
+        return ForSomeIn(tokens[1], tokens[2], tokens[3], location=locn)
+
+    raise ParseFatalException(string, location, "unexpected quantified expression")
 
 
 @fatalexceptions
-def parse_term(string: str, location: int, tokens: ParseResults) -> Expr:
-    if isinstance(tokens[0], ID):
-        return Variable(tokens[0], negative=False, location=tokens[0].location)
-    return tokens[0]
+def parse_comprehension(string: str, location: int, tokens: ParseResults) -> Comprehension:
+    tokens, locn = evaluate_located_expression(string, tokens)
+    condition = tokens[3] if len(tokens) > 3 else TRUE
+    return Comprehension(tokens[0], tokens[1], tokens[2], condition, location=locn)
 
 
 @fatalexceptions
-def parse_relation(string: str, location: int, tokens: ParseResults) -> Relation:
-    def locn() -> Location:
-        return Location(tokens[0].location.start, tokens[0].location.source, tokens[2].location.end)
-
-    if tokens[1] == "<":
-        return Less(tokens[0], tokens[2], locn())
-    if tokens[1] == "<=":
-        return LessEqual(tokens[0], tokens[2], locn())
-    if tokens[1] == "=":
-        return Equal(tokens[0], tokens[2], locn())
-    if tokens[1] == ">=":
-        return GreaterEqual(tokens[0], tokens[2], locn())
-    if tokens[1] == ">":
-        return Greater(tokens[0], tokens[2], locn())
-    if tokens[1] == "/=":
-        return NotEqual(tokens[0], tokens[2], locn())
-    raise ParseFatalException(string, location, "unexpected relation operator")
+def parse_relational_operator(string: str, location: int, tokens: ParseResults) -> Relation:
+    result: List[Relation] = tokens[0]
+    while len(result) > 1:
+        left = result.pop(0)
+        operator = result.pop(0)
+        right = result.pop(0)
+
+        assert left.location
+        assert right.location
+        assert left.location.source == right.location.source
+
+        locn = Location(left.location.start, left.location.source, right.location.end)
+
+        relation: Relation
+        if operator == "=":
+            relation = Equal(left, right, location=locn)
+        elif operator == "/=":
+            relation = NotEqual(left, right, location=locn)
+        elif operator == "<":
+            relation = Less(left, right, location=locn)
+        elif operator == "<=":
+            relation = LessEqual(left, right, location=locn)
+        elif operator == ">=":
+            relation = GreaterEqual(left, right, location=locn)
+        elif operator == ">":
+            relation = Greater(left, right, location=locn)
+        elif operator == "in":
+            relation = In(left, right, location=locn)
+        elif operator == "not in":
+            relation = NotIn(left, right, location=locn)
+        else:
+            raise ParseFatalException(string, location, "unexpected relation operator")
```

Comment:
I wonder if having a `relationdict: Dict[str, Type[Relation]]` and then calling it with `relation = relationdict[operator](left, right, location=locn)` would be more efficient and maybe a bit less repetitive.